src/events/SDL_events.c

```diff
@@ -887,9 +887,9 @@ SDL_WaitEventTimeout(SDL_Event * event, int timeout)
     case 0:
         break;
     default:
-        /* Check whether we have reached the end of the poll cycle */
+        /* Check whether we have reached the end of the poll cycle, and no more events are left */
         if (timeout == 0 && event && event->type == SDL_POLLSENTINEL) {
-            return 0;
+            return SDL_PeepEvents(event, 1, SDL_GETEVENT, SDL_FIRSTEVENT, SDL_LASTEVENT) == 1;
```

Comment:
After [searching through GitHub](https://github.com/search?l=C%2B%2B&p=1&q=SDL_PollEvent&type=Code) for uses of `SDL_PollEvent`, there is an alarming amount of strange/incorrect usage (mostly in beginner projects), such as:
* Only calling it once per frame
* Not checking the return value
* Calling `SDL_PollEvent(NULL)` in a loop
* Polling until a certain event (i.e a key press), and pushing the rest back into the queue

I'd say those first 3 issues are broken by design, but the last one (and similar uses of user-added events inside a poll cycle) could be an issue, so I decided it's better to err on the side of caution and ignore the sentinel if there are still more events in the queue.